Laba №5 Huffman Coding/HuffmanTree.cpp

```diff
@@ -0,0 +1,344 @@
+    printHorizontal(root->leftChild(), marginLeft + levelSpacing, levelSpacing);
+}
+
+HuffmanTree::~HuffmanTree()
+{
+}
+
+void HuffmanTree::clear()
+{
+    clear(m_root);
+    m_root = nullptr;
+}
+
+void HuffmanTree::print()
+{
+    m_root->printHorizontal(m_root, 0, 15);
+}
+
+void HuffmanTree::build(const std::string& textFilename)
+{
+    std::string text;
+
+    std::ifstream textFile;
+    textFile.open(textFilename);
+
+    if (!textFile.is_open())
+        return;
+
+    char ch;
+
+    textFile >> ch;
+
+    while (textFile.good())
+    {
+        text += ch;
+
+        textFile >> ch;
+    }
+
+    textFile.close();
+
+    BoolVector BV(256, 0);
+
+    if (m_root)
+    {
+        clear();
+    }
+
+    std::vector<Node*> nodes;
+    for (const char& symbol : text)
+    {
+        std::vector<Node*>::iterator it = std::find_if(
+            nodes.begin(),
+            nodes.end(),
+            [&symbol](Node* node) // Безымянная функция
+            {
+                return node->symbols().bitValue(symbol);
+            }
+        );
+
+        if (it == nodes.end())
+        {
+            nodes.push_back(new Node());
+
+            BoolVector BV = nodes.back()->symbols();
+
+            BV[symbol] = true;
+
+            nodes.back()->setSymbols(BV);
+        }
+        else
+        {
+            Node* node = *it;
+            node->setFrequency(node->frequency() + 1);
+        }
+    }
+
+    // Ещё одна безымянная функция, сложили в переменную
+    auto sorter = [](Node* a, Node* b) -> bool
+        {
+            return a->frequency() < b->frequency();
+        };
+    std::sort(nodes.begin(), nodes.end(), sorter);
+
+    while (nodes.size() > 1)
+    {
+        Node* root = new Node();
+        root->setFrequency(nodes[0]->frequency() + nodes[1]->frequency());
+
+        BoolVector BV = root->symbols();
+
+        for (int i = 0; i < root->symbols().length(); ++i)
+        {
+            BV[i] = nodes[0]->symbols()[i] || nodes[1]->symbols()[i];
+        }
+
+        root->setSymbols(BV);
+
+        root->setLeftChild(nodes[0]);
+        root->setRightChild(nodes[1]);
+
+        nodes.erase(nodes.begin(), nodes.begin() + 2);
+        nodes.push_back(root);
+        std::sort(nodes.begin(), nodes.end(), sorter);
+    }
+
+    m_root = nodes.front();
+}
+
+bool isEndFile(const std::string& fileBI);
+
+double HuffmanTree::encode(const std::string& textFilename, const std::string& encodedTextFilename)
+{
+    std::ifstream textFile;
+    std::ofstream encodedTextFile;
+
+    std::string encodedText;
+
+    int encodedTextSize = 0;
+    char symbol;
+
+    textFile.open(textFilename);
+    encodedTextFile.open(encodedTextFilename);
+
+    if (isEndFile(textFilename) || !textFile.is_open() || !encodedTextFile.is_open())
+    {
+        textFile.close();
+        encodedTextFile.close();
+        return -1;
+    }
+
+    if (m_root == nullptr)
+    {
+        build(textFilename);
+    }
+
+    textFile >> symbol;

```

Comment:
Т.к. `build` можно вызвать отдельно, она всё равно должна проверять файл на корректность, а следовательно проверку `isEndFile` (которая скорее `isEmptyFile`) можно убрать и просто заменить на проверку корня.

```suggestion
    if (m_root == nullptr)
    {
        build(textFilename);
    }
    
    textFile.open(textFilename);
    encodedTextFile.open(encodedTextFilename);

    if (!m_root || !textFile.is_open() || !encodedTextFile.is_open())
    {
        textFile.close();
        encodedTextFile.close();
        return -1;
    }

    textFile >> symbol;
```